wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ConversationListViewController+NavigationBar.swift

```diff
@@ -112,20 +127,118 @@ extension ConversationListViewController {
+        let allConversationsAction = createFilterAction(
+            title: FilterMenuLocale.AllConversations.title,
+            filter: .allConversations,
+            isSelected: selectedFilter == .allConversations
+        )
+        let favoritesAction = createFilterAction(
+            title: FilterMenuLocale.Favorites.title,
+            filter: .favorites,
+            isSelected: selectedFilter == .favorites
+        )
+        let groupsAction = createFilterAction(
+            title: FilterMenuLocale.Groups.title,
+            filter: .groups,
+            isSelected: selectedFilter == .groups
+        )
+        let oneToOneConversationsAction = createFilterAction(
+            title: FilterMenuLocale.OneOnOneConversations.title,
+            filter: .oneToOneConversations,
+            isSelected: selectedFilter == .oneToOneConversations
+        )
+
+        // Create the menu
+        let filterMenu = UIMenu(
+            children: [
+                allConversationsAction,
+                favoritesAction,
+                groupsAction,
+                oneToOneConversationsAction
+            ]
+        )
+
+        // Create the filter button and assign the menu
+        let filterButton = UIButton(type: .system)
+        filterButton.setImage(selectedFilterImage, for: .normal)
+        filterButton.showsMenuAsPrimaryAction = true
+        filterButton.menu = filterMenu
+
+        navigationItem.rightBarButtonItems?.append(UIBarButtonItem(customView: filterButton))
+    }
+
+    private func createFilterAction(
+        title: String,
+        filter: FilterType,
+        isSelected: Bool
+    ) -> UIAction {
+        let imageName = getFilterImageName(for: filter, isSelected: isSelected)
+
+        let font = UIFont.systemFont(ofSize: 17)
+        let configuration = UIImage.SymbolConfiguration(font: font)
+        let actionImage = UIImage(systemName: imageName)?.applyingSymbolConfiguration(configuration)
+
+        // Apply the tint color conditionally based on the selection state
+        let tintedActionImage: UIImage?
+        if isSelected {
+            tintedActionImage = actionImage?.withTintColor(UIColor.accent(), renderingMode: .alwaysOriginal)
+        } else {
+            tintedActionImage = actionImage
+        }
+
+        let titleAttributes: [NSAttributedString.Key: Any] = [
+            .foregroundColor: isSelected ? UIColor.accent() : SemanticColors.Label.textDefault
+        ]
+
+        let attributedTitle = NSAttributedString(string: title, attributes: titleAttributes)
+
+        let action = UIAction(title: title, image: tintedActionImage) { [weak self] _ in
+            self?.selectedFilter = filter
+            self?.setupRightNavigationBarButtons()
+        }
+
+        action.setValue(attributedTitle, forKey: "attributedTitle")
+
+        return action
+    }
+
+    private func getFilterImageName(for filter: FilterType, isSelected: Bool) -> String {
```

Comment:
I'm wondering if we could extract this and maybe more of the new code from the `ConversationListViewController`.

Reply:
I was thinking the same thing. But for now decided to keep it here. In a follow up PR we can separate it into its own object.